{
 "cells": [
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "***\n",
    "# Table of Contents\n",
    "***\n",
    "1. [LOH_pipeline_2.sh Program Versions](#program_versions)\n",
    "1. [Bash Pipeline](#bash)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "***\n",
    "# LOH_pipeline_2.sh Program Versions <a id=\"program_versions\"></a>\n",
    "***\n",
    "LOH_pipeline_2.sh\n",
    "```bash\n",
    "#paths to required programs\n",
    "BWA=\"/mnt/disk4/labs/salipante/programs/bwa-0.7.12/bwa\"\n",
    "SAMTOOLS=\"/mnt/disk4/labs/salipante/programs/samtools-0.1.19/samtools\"\n",
    "SEQUENZAUTILS=\"/home/local/AMC/nithisha/R/R.3.3-library/sequenza/exec/sequenza-utils.py\"\n",
    "ALIGNMENT_REF_GENOME=\"/mnt/disk2/com/Genomes/hg19_chr/human_g1k_v37.fasta\"  \n",
    "GC_WINDOW=\"/mnt/disk4/labs/salipante/stevesal/LOH_calling/scripts/gcwindow/hg19.gc5Base_v4.txt.gz\"\n",
    "SUPER_DEDUPER=\"/mnt/disk4/labs/salipante/programs/Super-Deduper-master/super_deduper\"\n",
    "#SCORING_SCRIPT=\"/mnt/disk4/labs/salipante/stevesal/LOH_calling/scripts/sequenza_loss_score_v2.pl\"  \n",
    "SCORING_SCRIPT_PYTHON=\"/mnt/disk4/labs/salipante/nithisha/LOH_Scoring/nit_code/LOH_score.py\"\n",
    "```\n",
    "\n",
    "Program Versions\n",
    "\n",
    "| tool/resource | version used | latest | comments | link |\n",
    "| ------------- | ------------ | ------ | -------- | ---- |\n",
    "| Super DeDuper |-|end of life|mpileup (duplicate removal)| https://github.com/dstreett/Super-Deduper<br>https://github.com/ibest/HTStream |\n",
    "| bwa           |0.7.12|0.7.17|mpileup (align, sampe)| https://sourceforge.net/projects/bio-bwa/files/ |\n",
    "| samtools      |0.1.19|1.10.2|mpileup (sort, index)| https://github.com/samtools/samtools |\n",
    "| sequenza utils |3.0.0|3.0.0|sequenza| https://bitbucket.org/sequenzatools/sequenza-utils/src/master/<br>https://sequenza-utils.readthedocs.io/en/latest/index.html |\n",
    "| scoring script |-|-|loss scoring|local|\n",
    "| alignment ref genome |37|-|-|-|\n",
    "| GC window     |-|-|-|-|\n"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "***\n",
    "# Bash Pipeline <a id=\"bash\"></a>\n",
    "***\n",
    "  \n",
    "```bash\n",
    "export R_LIBS=\"/home/local/AMC/nithisha/R/R.3.3-library\"\n",
    "set -x -e\n",
    "#declare arguments\n",
    "TUMOR_PFX=$1 #tumor prefix\n",
    "NORMAL_PFX=$2\n",
    "SAVEROOT=$3 #path to where you want your results saved\n",
    "DATAPATH=$4 #path to directory containing sample-specific fastq files\n",
    "CORES=$5\n",
    "```\n",
    "  \n",
    "```bash\n",
    "#paths to required programs\n",
    "BWA=\"/mnt/disk4/labs/salipante/programs/bwa-0.7.12/bwa\"\n",
    "SAMTOOLS=\"/mnt/disk4/labs/salipante/programs/samtools-0.1.19/samtools\"\n",
    "SEQUENZAUTILS=\"/home/local/AMC/nithisha/R/R.3.3-library/sequenza/exec/sequenza-utils.py\"\n",
    "ALIGNMENT_REF_GENOME=\"/mnt/disk2/com/Genomes/hg19_chr/human_g1k_v37.fasta\"  \n",
    "GC_WINDOW=\"/mnt/disk4/labs/salipante/stevesal/LOH_calling/scripts/gcwindow/hg19.gc5Base_v4.txt.gz\"\n",
    "SUPER_DEDUPER=\"/mnt/disk4/labs/salipante/programs/Super-Deduper-master/super_deduper\"\n",
    "#SCORING_SCRIPT=\"/mnt/disk4/labs/salipante/stevesal/LOH_calling/scripts/sequenza_loss_score_v2.pl\"  \n",
    "SCORING_SCRIPT_PYTHON=\"/mnt/disk4/labs/salipante/nithisha/LOH_Scoring/nit_code/LOH_score.py\"\n",
    "#SUMMARY_FILE=\"/mnt/disk4/labs/salipante/nithisha/LOH_Scoring/nit_summary_file.txt\"\n",
    "\n",
    "#create a subdirectory for this tumor specimen\n",
    "SAVEPATH=$SAVEROOT/$TUMOR_PFX\n",
    "mkdir -p $SAVEPATH\n",
    "\n",
    "#find tumor sequences\n",
    "TUMOR_R1=$(ls $DATAPATH/$TUMOR_PFX/*_1_sequence.txt.gz)\n",
    "TUMOR_R2=$(ls $DATAPATH/$TUMOR_PFX/*_2_sequence.txt.gz)\n",
    "\n",
    "#find normal sequences\n",
    "NORMAL_R1=$(ls $DATAPATH/$NORMAL_PFX/*_1_sequence.txt.gz)\n",
    "NORMAL_R2=$(ls $DATAPATH/$NORMAL_PFX/*_2_sequence.txt.gz)\n",
    "```\n",
    "  \n",
    "#### START: mpileup stage  \n",
    "```bash\n",
    "#-----------------------------------------------------------------------------------------------\n",
    "#CHECK FOR MPILEUPS, create them if they dont exist\n",
    "#-----------------------------------------------------------------------------------------------\n",
    "#ls $SAVEPATH/$NORMAL_PFX.mpileup.gz\n",
    "if [ ! -f $SAVEPATH/$NORMAL_PFX.mpileup.gz ] #remember spaces in []\n",
    "then \n",
    "```\n",
    "  \n",
    "Remove tumor duplicates, gzip fastqs  \n",
    "```bash\n",
    "#-----------------------------------------------------------------------------------------------\n",
    "#preprocess tumor reads \n",
    "$SUPER_DEDUPER -1 $TUMOR_R1 -2 $TUMOR_R2 -p $SAVEPATH/$TUMOR_PFX\n",
    "echo \"line 41...Super deduper for tumor done, check for 2 _nodup_PE1/2.fastq files\"\n",
    "gzip $SAVEPATH/${TUMOR_PFX}_nodup_PE1.fastq &\n",
    "gzip $SAVEPATH/${TUMOR_PFX}_nodup_PE2.fastq &\n",
    "\n",
    "wait\n",
    "echo \"line 46...gzipping tumor done, check for 2 _nodup_PE1/2.fastq.gz files\"\n",
    "```\n",
    "***\n",
    "\n",
    "BWA/SAMTOOLS align, sampe, sort, index (tumor)  \n",
    "```bash\n",
    "#align tumor\n",
    "#Align fastq files to gatk reference genome - this finds SA coordinates of input reads (.sai- suffix array indices)\n",
    "$BWA aln -t $CORES $ALIGNMENT_REF_GENOME $SAVEPATH/${TUMOR_PFX}_nodup_PE1.fastq.gz > $SAVEPATH/$TUMOR_PFX.1.sai 2>> $SAVEPATH/$TUMOR_PFX.log\n",
    "$BWA aln -t $CORES $ALIGNMENT_REF_GENOME $SAVEPATH/${TUMOR_PFX}_nodup_PE2.fastq.gz > $SAVEPATH/$TUMOR_PFX.2.sai 2>> $SAVEPATH/$TUMOR_PFX.log\n",
    "echo \"line 52...BWA aln for tumors done, check for 2 .sai files\"\n",
    "\n",
    "# .sai to .sam file and convert SA coordinates to chromosomal loci\n",
    "$BWA sampe -r \"@RG\\tID:${PFX}\\tPL:ILLUMINA\\tPU:NA\\tLB:null\\tSM:${TUMOR_PFX:?}\" $ALIGNMENT_REF_GENOME $SAVEPATH/$TUMOR_PFX.1.sai $SAVEPATH/$TUMOR_PFX.2.sai $SAVEPATH/${TUMOR_PFX}_nodup_PE1.fastq.gz $SAVEPATH/${TUMOR_PFX}_nodup_PE2.fastq.gz 2>> $SAVEPATH/$TUMOR_PFX.log >$SAVEPATH/$TUMOR_PFX.sam\n",
    "echo \"line 56... BWA sampe done for tumor, check for 1 .sam file\"\n",
    "#remove large files\n",
    "rm -f $SAVEPATH/${TUMOR_PFX}_nodup_PE1.fastq.gz\n",
    "rm -f $SAVEPATH/${TUMOR_PFX}_nodup_PE2.fastq.gz\n",
    "\n",
    "#use SAMTOOLS to view, sort and index, .sam (sequence alignment/mapping) to .bam\n",
    "#view - converts .sam to .bam without any action taken\n",
    "$SAMTOOLS view -u -b -S -F 4 -Q 20 $SAVEPATH/$TUMOR_PFX.sam | $SAMTOOLS sort - $SAVEPATH/$TUMOR_PFX 2>> $SAVEPATH/$TUMOR_PFX.log\n",
    "# u: uncompressed if pipe is used, b: bam file output, S: checking for compatibility with older SAMTOOLS version\n",
    "# F: do not output alignments with FLAG, q: skip alignments with MAPQ lesser than\n",
    "# T: prefix to be added to output\n",
    "\n",
    "#index BAM file (Index a coordinate-sorted BAM file for fast random access)\n",
    "$SAMTOOLS index $SAVEPATH/$TUMOR_PFX.bam\n",
    "echo \"line 70... SAMTOOLS view, sort and index done for tumor, check for 1 .bam file\"\n",
    "#remove more intermediate files\n",
    "rm -f $SAVEPATH/$TUMOR_PFX.1.sai\n",
    "rm -f $SAVEPATH/$TUMOR_PFX.2.sai\n",
    "rm -f $SAVEPATH/$TUMOR_PFX.sam\n",
    "```\n",
    "  \n",
    "Remove normal duplicates, gzip fastqs  \n",
    "```bash\n",
    "#-----------------------------------------------------------------------------------------------\n",
    "#preprocess normal reads \n",
    "$SUPER_DEDUPER -1 $NORMAL_R1 -2 $NORMAL_R2 -p $SAVEPATH/$NORMAL_PFX\n",
    "\n",
    "gzip $SAVEPATH/${NORMAL_PFX}_nodup_PE1.fastq &\n",
    "gzip $SAVEPATH/${NORMAL_PFX}_nodup_PE2.fastq &\n",
    "\n",
    "wait\n",
    "```\n",
    "  \n",
    "BWA/SAMTOOLS align, sampe, sort, index (normal)  \n",
    "```bash\n",
    "#align normal\n",
    "#Align fastq files to gatk reference genome - this finds SA coordinates of input reads (.sai- suffix array indices)\n",
    "$BWA aln -t $CORES $ALIGNMENT_REF_GENOME $SAVEPATH/${NORMAL_PFX}_nodup_PE1.fastq.gz > $SAVEPATH/$NORMAL_PFX.1.sai 2>> $SAVEPATH/$NORMAL_PFX.log\n",
    "$BWA aln -t $CORES $ALIGNMENT_REF_GENOME $SAVEPATH/${NORMAL_PFX}_nodup_PE2.fastq.gz > $SAVEPATH/$NORMAL_PFX.2.sai 2>> $SAVEPATH/$NORMAL_PFX.log\n",
    "\n",
    "# .sai to .sam file and convert SA coordinates to chromosomal loci\n",
    "$BWA sampe -r \"@RG\\tID:${PFX}\\tPL:ILLUMINA\\tPU:NA\\tLB:null\\tSM:${NORMAL_PFX:?}\" $ALIGNMENT_REF_GENOME $SAVEPATH/$NORMAL_PFX.1.sai $SAVEPATH/$NORMAL_PFX.2.sai $SAVEPATH/${NORMAL_PFX}_nodup_PE1.fastq.gz $SAVEPATH/${NORMAL_PFX}_nodup_PE2.fastq.gz 2>>$SAVEPATH/$NORMAL_PFX.log > $SAVEPATH/$NORMAL_PFX.sam\n",
    "echo \"line 92... BWA done for normal, check for .sam file\"\n",
    "#remove large files\n",
    "rm -f $SAVEPATH/${NORMAL_PFX}_nodup_PE1.fastq.gz\n",
    "rm -f $SAVEPATH/${NORMAL_PFX}_nodup_PE2.fastq.gz\n",
    "\n",
    "#use SAMTOOLS to view, sort and index, .sam (sequence alignment/mapping) to .bam\n",
    "#view - converts .sam to .bam without any action taken\n",
    "\n",
    "$SAMTOOLS view -u -b -S -F 4 -q 20 $SAVEPATH/$NORMAL_PFX.sam | $SAMTOOLS sort - $SAVEPATH/$NORMAL_PFX 2>>$SAVEPATH/$NORMAL_PFX.log\n",
    "# u: uncompressed if pipe is used, b: bam file output, S: checking for compatibility with older SAMTOOLS version\n",
    "# F: do not output alignments with FLAG, q: skip alignments with MAPQ lesser than\n",
    "# T: prefix to be added to output\n",
    "echo \"line 104... SAMTOOLS view and sort done for normal, check for .bam file\"\n",
    "\n",
    "#index BAM file (Index a coordinate-sorted BAM file for fast random access)\n",
    "$SAMTOOLS index $SAVEPATH/$NORMAL_PFX.bam\n",
    "echo \"line 107... SAMTOOLS index done for normal, check for .bam file\"\n",
    "#remove more intermediate files\n",
    "rm -f $SAVEPATH/$NORMAL_PFX.1.sai\n",
    "rm -f $SAVEPATH/$NORMAL_PFX.2.sai\n",
    "rm -f $SAVEPATH/$NORMAL_PFX.sam\n",
    "```\n",
    "  \n",
    "```bash\n",
    "#-----------------------------------------------------------------------------------------------\n",
    "#CHECK for 2 bams at this stage, if there was a samtool permission denied, segmentation faults (core dumped) error, you will not have 2 bams - one for tumor and one for germline\n",
    "#at this point just exit and re-run this sample\n",
    "if [ ! -f $SAVEPATH/$TUMOR_PFX.bam ] && [ ! -f $SAVEPATH/$NORMAL_PFX.bam ]\n",
    "then\n",
    "echo 'ERROR ENCOUNTERED- RE_RUN SAMPLE'\n",
    "exit\n",
    "fi\n",
    "```\n",
    "  \n",
    "SAMTOOLS mpileup  \n",
    "```bash\n",
    "#-----------------------------------------------------------------------------------------------\n",
    "#Make mpileups and gzip them using SAMTOOLS\n",
    "$SAMTOOLS mpileup -f $ALIGNMENT_REF_GENOME -d 10000 -A -B $SAVEPATH/$TUMOR_PFX.bam | gzip > $SAVEPATH/$TUMOR_PFX.mpileup.gz\n",
    "$SAMTOOLS mpileup -f $ALIGNMENT_REF_GENOME -d 10000 -A -B $SAVEPATH/$NORMAL_PFX.bam | gzip > $SAVEPATH/$NORMAL_PFX.mpileup.gz\n",
    "#f: reference genome, d: max depth, A: Do not skip anomalous read pairs in variant calling\n",
    "#B: Disable probabilistic realignment for the computation of base alignment quality (BAQ). \n",
    "#BAQ is the Phred-scaled probability of a read base being misaligned. \n",
    "#Applying this option greatly helps to reduce false SNPs caused by misalignments. \n",
    "\n",
    "fi\n",
    "```\n",
    "  \n",
    "END: mpileup stage  \n",
    "#### START: Sequenza  \n",
    "Parse information from tumor and normal sequencing data (mpileup) to produce a file format suitable for the sequenza R package.  \n",
    "```bash\n",
    "#-----------------------------------------------------------------------------------------------\n",
    "#RUN SEQUENZA, PYTHON\n",
    "#-----------------------------------------------------------------------------------------------\n",
    "if [ ! -f $SAVEPATH/$TUMOR_PFX.binned.seqz.gz ]\n",
    "then\n",
    "python $SEQUENZAUTILS pileup2seqz -gc $GC_WINDOW -n $SAVEPATH/$NORMAL_PFX.mpileup.gz -t $SAVEPATH/$TUMOR_PFX.mpileup.gz | gzip > $SAVEPATH/$TUMOR_PFX.seqz.gz\n",
    "python $SEQUENZAUTILS seqz-binning -w 50 -s $SAVEPATH/$TUMOR_PFX.seqz.gz | gzip > $SAVEPATH/$TUMOR_PFX.binned.seqz.gz\n",
    "#gc: gc window, n: normal mpileup, t:tumor mpileup\n",
    "#w:-w indicate a window size (in bases), to be used for the binning, s:save to file??\n",
    "fi\n",
    "echo \"line 133... Sequenza in python done!\"\n",
    "```\n",
    "  \n",
    "sequenza R package  \n",
    "```bash\n",
    "#-----------------------------------------------------------------------------------------------\n",
    "#RUN SEQUENZA, R\n",
    "#-----------------------------------------------------------------------------------------------\n",
    "#Create an executable R script, run it and quit it!\n",
    "echo \"library(\\\"sequenza\\\")\">$SAVEPATH/$TUMOR_PFX.sequenza.r\n",
    "echo \"data.file <- \\\"$SAVEPATH/$TUMOR_PFX.binned.seqz.gz\\\"\" >> $SAVEPATH/$TUMOR_PFX.sequenza.r\n",
    "#echo \"data.file <- \\\"$SAVEPATH/$TUMOR_PFX.seqz.gz\\\"\" >> $SAVEPATH/$TUMOR_PFX.sequenza.r\n",
    "echo \"seqz.data <- read.seqz(data.file)\" >> $SAVEPATH/$TUMOR_PFX.sequenza.r\n",
    "echo \"gc.stats <- gc.sample.stats(data.file)\" >> $SAVEPATH/$TUMOR_PFX.sequenza.r\n",
    "echo \"test <- sequenza.extract(data.file)\" >> $SAVEPATH/$TUMOR_PFX.sequenza.r\n",
    "echo \"CP.example <- sequenza.fit(test)\" >> $SAVEPATH/$TUMOR_PFX.sequenza.r\n",
    "echo \"sequenza.results(sequenza.extract = test, cp.table = CP.example, sample.id = \\\"$TUMOR_PFX\\\", out.dir=\\\"$SAVEPATH\\\")\" >> $SAVEPATH/$TUMOR_PFX.sequenza.r\n",
    "echo \"cint <- get.ci(CP.example)\" >> $SAVEPATH/$TUMOR_PFX.sequenza.r\n",
    "\n",
    "#Plot cellularity\n",
    "echo \"jpeg(\\\"$SAVEPATH/$TUMOR_PFX.nitz.cellularity.jpg\\\")\" >> $SAVEPATH/$TUMOR_PFX.sequenza.r\n",
    "echo \"cp.plot(CP.example)\" >> $SAVEPATH/$TUMOR_PFX.sequenza.r\n",
    "echo \"cp.plot.contours(CP.example, add = TRUE, likThresh=c(0.95))\" >> $SAVEPATH/$TUMOR_PFX.sequenza.r\n",
    "echo \"dev.off()\" >> $SAVEPATH/$TUMOR_PFX.sequenza.r\n",
    "\n",
    "#Call CNVs\n",
    "echo \"cellularity <- cint\\$max.cellularity\" >> $SAVEPATH/$TUMOR_PFX.sequenza.r\n",
    "echo \"ploidy <- cint\\$max.ploidy\" >> $SAVEPATH/$TUMOR_PFX.sequenza.r\n",
    "echo \"avg.depth.ratio <- mean(test\\$gc\\$adj[,2])\" >> $SAVEPATH/$TUMOR_PFX.sequenza.r\n",
    "\n",
    "#Save parameters to file to file\n",
    "echo \"cellularity\" >> $SAVEPATH/$TUMOR_PFX.sequenza.r\n",
    "echo \"write(cellularity, file = \\\"$SAVEPATH/$TUMOR_PFX.nitz.cellularity.txt\\\")\" >> $SAVEPATH/$TUMOR_PFX.sequenza.r\n",
    "echo \"write(ploidy, file = \\\"$SAVEPATH/$TUMOR_PFX.nitz.ploidy.txt\\\")\" >>$SAVEPATH/$TUMOR_PFX.sequenza.r\n",
    "echo \"write(avg.depth.ratio, file = \\\"$SAVEPATH/$TUMOR_PFX.nitz.ave_depth.txt\\\")\" >> $SAVEPATH/$TUMOR_PFX.sequenza.r\n",
    "\n",
    "#Detect variant alleles\n",
    "echo \"mut.tab <- na.exclude(do.call(rbind, test\\$mutations))\" >> $SAVEPATH/$TUMOR_PFX.sequenza.r\n",
    "echo \"mut.alleles <- mufreq.bayes(mufreq = mut.tab\\$F, depth.ratio = mut.tab\\$adjusted.ratio, cellularity = cellularity, ploidy = ploidy, avg.depth.ratio = avg.depth.ratio)\" >> $SAVEPATH/$TUMOR_PFX.sequenza.r\n",
    "\n",
    "#Detect CN variation\n",
    "echo \"seg.tab <- na.exclude(do.call(rbind, test\\$segments))\" >> $SAVEPATH/$TUMOR_PFX.sequenza.r\n",
    "echo \"cn.alleles <- baf.bayes(Bf = seg.tab\\$Bf, depth.ratio = seg.tab\\$depth.ratio, cellularity = cellularity, ploidy = ploidy, avg.depth.ratio = avg.depth.ratio)\" >> $SAVEPATH/$TUMOR_PFX.sequenza.r\n",
    "echo \"seg.tab <- cbind(seg.tab, cn.alleles)\" >>$SAVEPATH/$TUMOR_PFX.sequenza.r\n",
    "echo \"seg.tab\" >> $SAVEPATH/$TUMOR_PFX.sequenza.r\n",
    "\n",
    "#write sequenza matrix to file, this will serve as input to loss score script's 2nd arg\n",
    "echo \"write.table(seg.tab, file = \\\"$SAVEPATH/$TUMOR_PFX.nitz.copynumber_calls.txt\\\", append = FALSE)\" >> $SAVEPATH/$TUMOR_PFX.sequenza.r\n",
    "\n",
    "#exit\n",
    "echo \"q()\" >> $SAVEPATH/$TUMOR_PFX.sequenza.r\n",
    "echo \"n\" >> $SAVEPATH/$TUMOR_PFX.sequenza.r\n",
    "\n",
    "#execute the R script\n",
    "R --vanilla < $SAVEPATH/$TUMOR_PFX.sequenza.r\n",
    "```\n",
    "  \n",
    "END: Sequenza  \n",
    "#### START: Scoring Script  \n",
    "```bash\n",
    "#Calculate loss score\n",
    "if [ -f $SAVEPATH/$TUMOR_PFX.nitz.copynumber_calls.txt ]\n",
    "then \n",
    "python $SCORING_SCRIPT_PYTHON $SAVEPATH/$TUMOR_PFX.nitz.copynumber_calls.txt $SAVEPATH/$TUMOR_PFX.nitz.score.txt 0.75\n",
    "\n",
    "#Append some usefull QC factoids to the lossScore output text\n",
    "echo \"\" >> $SAVEPATH/$TUMOR_PFX.nitz.score.txt\n",
    "echo -n \"Estimated tumor cellularity: \" >> $SAVEPATH/$TUMOR_PFX.nitz.score.txt\n",
    "cat $SAVEPATH/$TUMOR_PFX.nitz.cellularity.txt >> $SAVEPATH/$TUMOR_PFX.nitz.score.txt\n",
    "echo -n \"Estimated ploidy: \" >> $SAVEPATH/$TUMOR_PFX.nitz.score.txt\n",
    "cat $SAVEPATH/$TUMOR_PFX.nitz.ploidy.txt >> $SAVEPATH/$TUMOR_PFX.nitz.score.txt\n",
    "\n",
    "else\n",
    "exit\n",
    "\n",
    "fi\n",
    "```\n",
    "END: Scoring Script  "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.6.7"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 2
}
